p2p/chain.py

```diff
@@ -437,19 +465,60 @@ def __init__(self,
 
         return dict((part.unique_key, part.part) for part in parts)
 
+    def _get_throughput_stats(self, peer: BasePeer, body_part: BlockPartEnum) -> ThroughputTracker:
+        return self._throughput_stats[body_part][peer]
+
+    def deregister_peer(self, peer: BasePeer) -> None:
+        # Delete all in-progress stats for the given peer. Otherwise, if a peer disconnects before
+        # sending a block part, stats would fail when starting to track throughput the 2nd time.
+        # Also, we don't want a perpetually growing dict of stats.
+        for part in BlockPartEnum:
+            del self._throughput_stats[part][peer]
```

Comment:
Could this raise a KeyError if the peer never sent us anything?

Reply:
Ah yes, a bug in my mental representation of `defaultdict`